xapian-core/cluster/cluster.cc

```diff
@@ -324,91 +349,189 @@ Point
 Cluster::get_index(unsigned int index) const
 {
     LOGCALL(API, Point, "Cluster::get_index", index);
+    return internal->get_index(index);
+}
+
+Point
+Cluster::Internal::get_index(unsigned int index) const
+{
     return cluster_docs[index];
 }
 
+void
+ClusterSet::operator=(const ClusterSet &other)
+{
+    // pointers are reference counted.
+    internal = other.internal;
+}
+
+ClusterSet::ClusterSet(const ClusterSet &other)
+	: internal(other.internal)
+{
+}
+
+ClusterSet::ClusterSet() : internal(new Xapian::ClusterSet::Internal)
+{
+}
+
+ClusterSet::~ClusterSet()
+{
+}
+
+doccount
+ClusterSet::Internal::size() const
+{
+    return clusters.size();
+}
+
 doccount
 ClusterSet::size() const
 {
     LOGCALL(API, doccount, "ClusterSet::size", NO_ARGS);
-    return clusters.size();
+    return internal->size();
 }
 
 void
-ClusterSet::add_cluster(Cluster &c)
+ClusterSet::Internal::add_cluster(const Cluster &c)
 {
-    LOGCALL_VOID(API, "ClusterSet::add_cluster", c);
     clusters.push_back(c);
 }
 
+void
+ClusterSet::add_cluster(const Cluster &c)
+{
+    LOGCALL_VOID(API, "ClusterSet::add_cluster", c);
+    internal->add_cluster(c);
+}
+
 Cluster&
-ClusterSet::operator[](doccount i)
+ClusterSet::Internal::get_cluster(doccount i)
 {
     return clusters[i];
 }
 
+Cluster&
+ClusterSet::operator[](doccount i)
+{
+    return internal->get_cluster(i);
+}
+
+void
+ClusterSet::Internal::add_to_cluster(const Point &x, unsigned int i)
+{
+    clusters[i].add_point(x);
+}
+
 void
 ClusterSet::add_to_cluster(const Point &x, unsigned int i)
 {
    LOGCALL_VOID(API, "ClusterSet::add_to_cluster", x | i);
-   clusters[i].add_point(x);
+   internal->add_to_cluster(x, i);
 }
 
 void
-ClusterSet::recalculate_centroids()
+ClusterSet::Internal::recalculate_centroids()
 {
-    LOGCALL_VOID(API, "ClusterSet::recalculate_centroids", NO_ARGS);
     for (vector<Cluster>::iterator it = clusters.begin(); it != clusters.end(); ++it)
 	(*it).recalculate();
 }
 
+void
+ClusterSet::recalculate_centroids()
+{
+    LOGCALL_VOID(API, "ClusterSet::recalculate_centroids", NO_ARGS);
+    internal->recalculate_centroids();
+}
+
+void
+ClusterSet::clear_clusters()
+{
+    LOGCALL_VOID(API, "ClusterSet::clear_clusters", NO_ARGS);
+    internal->clear_clusters();
+}
+
+void
+ClusterSet::Internal::clear_clusters()
+{
+    for (vector<Cluster>::iterator it = clusters.begin(); it !=clusters.end(); ++it)
```

Comment:
I'd recommend using C++11 `for (x : c)` instead of an explicit loop between `c.begin()` and `c.end()`, but as-is there's a missing space after `!=`.